plugins/audio-s5gen2/fu-audio-s5gen2-hid-device.c

```diff
@@ -0,0 +1,291 @@
+		return FALSE;
+
+	msg = fu_struct_qc_hid_response_parse(buf, FU_STRUCT_QC_HID_RESPONSE_SIZE, 0, error);
+	if (msg == NULL)
+		return FALSE;
+
+	if (!fu_memcpy_safe(data_in,
+			    data_len,
+			    0,
+			    msg->data,
+			    msg->len,
+			    FU_STRUCT_QC_HID_RESPONSE_OFFSET_PAYLOAD,
+			    fu_struct_qc_hid_response_get_payload_len(msg),
+			    error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_hid_device_msg_cmd(FuDevice *device, guint8 *data, gsize data_len, GError **error)
+{
+	FuQcS5gen2HidDevice *self = FU_QC_S5GEN2_HID_DEVICE(device);
+	g_autoptr(GByteArray) msg = fu_struct_qc_hid_command_new();
+
+	fu_struct_qc_hid_command_set_payload_len(msg, data_len);
+	if (!fu_struct_qc_hid_command_set_payload(msg, data, data_len, error))
+		return FALSE;
+
+	return fu_hid_device_set_report(FU_HID_DEVICE(self),
+					0x03,
+					msg->data,
+					FU_STRUCT_QC_HID_COMMAND_SIZE,
+					0,
+					FU_HID_DEVICE_FLAG_IS_FEATURE,
+					error);
+}
+
+static gboolean
+fu_qc_s5gen2_hid_device_probe(FuDevice *device, GError **error)
+{
+	FuHidDevice *hid_device = FU_HID_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	GUsbInterface *iface = NULL;
+	g_autoptr(GPtrArray) ifaces = NULL;
+
+	ifaces = g_usb_device_get_interfaces(usb_device, error);
+	if (ifaces == NULL)
+		return FALSE;
+	/* need the second HID interface */
+
+	if (ifaces->len <= HID_IFACE) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "transitional device detected");
+		return FALSE;
+	}
+
+	iface = g_ptr_array_index(ifaces, HID_IFACE);
+	if (g_usb_interface_get_class(iface) != G_USB_DEVICE_CLASS_HID) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "target interface is not HID");
+		return FALSE;
+	}
+
+	fu_hid_device_set_interface(hid_device, HID_IFACE);
+	fu_hid_device_set_ep_addr_in(hid_device, HID_EP_IN);
+	fu_hid_device_set_ep_addr_out(hid_device, HID_EP_OUT);
+
+	/* FuHidDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_qc_s5gen2_hid_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	fu_device_set_physical_id(device, "hid");
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_hid_device_setup(FuDevice *device, GError **error)
+{
+	FuQcS5gen2HidDevice *self = FU_QC_S5GEN2_HID_DEVICE(device);
+	FuQcS5gen2DeviceClass *klass_core;
+
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_qc_s5gen2_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	self->core = fu_qc_s5gen2_device_new(device);
+	if (self->core == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "unable to create the core");
+		return FALSE;
+	}
+	klass_core = FU_QC_S5GEN2_DEVICE_GET_CLASS(self->core);
+	klass_core->msg_out = fu_qc_s5gen2_hid_device_msg_out;
+	klass_core->msg_in = fu_qc_s5gen2_hid_device_msg_in;
+	klass_core->msg_cmd = fu_qc_s5gen2_hid_device_msg_cmd;
```

Comment:
I don't think this is right; we're setting the interface for this specific instance, not all instances of FuQcS5gen2Device -- e.g. we're being more permissive even than an _interface_. If you agree we can move the vfuncs into FuQcS5gen2DevicePrivate or FuQcS5gen2Device rather than the class struct.

Reply:
https://github.com/fwupd/fwupd/pull/6900 is my idea -- I've layered this branch on top of that and it makes it a lot cleaner.